catch/unit/math/math_common.hh

```diff
@@ -0,0 +1,217 @@
+        const auto actual_val = y_.ptr()[i];
+        const auto ref_val = static_cast<T>(ref_func(xss[i]...));
+        const auto validator = validator_builder(ref_val, xss[i]...);
+
+        if (!validator->match(actual_val)) {
+          const auto log = MakeLogMessage(actual_val, xss[i]...) + validator->describe() + "\n";
+          INFO(log);
+          REQUIRE(false);
+        }
+      }
+
+      return;
+    }
+
+    const auto task = [&, this](size_t iters, size_t base_idx) {
+      for (auto i = 0u; i < iters; ++i) {
+        if (fail_flag_.load(std::memory_order_relaxed)) return;
+
+        const auto actual_val = y_.ptr()[base_idx + i];
+        const auto ref_val = static_cast<T>(ref_func(xss[base_idx + i]...));
+        const auto validator = validator_builder(ref_val, xss[base_idx + i]...);
+
+        if (!validator->match(actual_val)) {
+          fail_flag_.store(true, std::memory_order_relaxed);
+          // Several threads might have passed the first check, but failed validation. On the
+          // chance of this happening, access to the string stream must be serialized.
+          const auto log =
+              MakeLogMessage(actual_val, xss[base_idx + i]...) + validator->describe() + "\n";
+          {
+            std::lock_guard lg{mtx_};
+            error_info_ += log;
+          }
+          return;
+        }
+      }
+    };
+
+    const auto task_count = thread_pool.thread_count();
+    const auto chunk_size = num_args / task_count;
+    const auto tail = num_args % task_count;
+
+    auto base_idx = 0u;
+    for (auto i = 0u; i < task_count; ++i) {
+      const auto iters = chunk_size + (i < tail);
+      thread_pool.Post([=, &task] { task(iters, base_idx); });
+      base_idx += iters;
+    }
+
+    thread_pool.Wait();
+
+    INFO(error_info_);
+    REQUIRE(!fail_flag_);
+  }
+
+  template <typename... Args> std::string MakeLogMessage(T actual_val, Args... args) {
+    std::stringstream ss;
+    ss << "Input value(s): " << std::scientific
+       << std::setprecision(std::numeric_limits<T>::max_digits10 - 1);
+    ((ss << " " << args), ...) << "\n" << actual_val << " ";
+
+    return ss.str();
+  }
+};
+
+template <typename T> struct RefType {};
+
+template <> struct RefType<float> { using type = double; };
+
+template <> struct RefType<double> { using type = long double; };
+
+template <typename T> using RefType_t = typename RefType<T>::type;
+
+template <typename F> auto GetOccupancyMaxPotentialBlockSize(F kernel) {
+  int grid_size = 0, block_size = 0;
+  HIP_CHECK(hipOccupancyMaxPotentialBlockSize(&grid_size, &block_size, kernel, 0, 0));
+  return std::make_tuple(grid_size, block_size);
+}
+
+inline size_t GetMaxAllowedDeviceMemoryUsage() {
+  hipDeviceProp_t props;
+  HIP_CHECK(hipGetDeviceProperties(&props, 0));
+  return props.totalGlobalMem * (cmd_options.accuracy_max_memory * 0.01f);
+}
+
+inline uint64_t GetTestIterationCount() { return cmd_options.accuracy_iterations; }
```

Comment:
Can we consider the type and arity here?   There is no need to do 2^32 iterations for a half precision unary function.

Reply:
See [comment](https://github.com/ROCm-Developer-Tools/hip-tests/pull/233#discussion_r1193570224)